File: android/app/src/main/kotlin/com/ekodide/android/core/Lacre.kt
package com.ekodide.android.core

import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec

/**
 * O lacre: assina/verifica cada mensagem com o segredo (HMAC-SHA256). O segredo
 * NUNCA trafega. Espelho byte-idêntico do lacre.py.
 *
 * Esta parte cobre o lado de PRODUÇÃO (assinar/empacotar), provável de testar no
 * JVM contra os vetores-ouro. A verificação (desempacotar) precisa de um parser
 * JSON e entra junto com o servidor.
 */
object Lacre {
    private const val HEX = "0123456789abcdef"

    /** Janela do carimbo de tempo (segundos): barra repetição de mensagens antigas. */
    const val JANELA_SEGUNDOS = 300L

    /** HMAC-SHA256 de `corpo` com `segredo`, em hex minúsculo (igual ao .hexdigest()). */
    fun assinar(corpo: ByteArray, segredo: String): String {
        val mac = Mac.getInstance("HmacSHA256")
        mac.init(SecretKeySpec(segredo.toByteArray(Charsets.UTF_8), "HmacSHA256"))
        return mac.doFinal(corpo).toHexLower()
    }

    /**
     * Carimba o tempo, assina e devolve os bytes prontos pra enviar. `agora` em
     * segundos epoch (default = relógio atual). Espelha lacre.empacotar.
     */
    fun empacotar(carga: Map<String, Any?>, segredo: String, agora: Long? = null): ByteArray {
        val ts = agora ?: (System.currentTimeMillis() / 1000)
        val selada = LinkedHashMap<String, Any?>(carga).apply { put("ts", ts) }
        val assinatura = assinar(CanonicalJson.encode(selada).toByteArray(Charsets.UTF_8), segredo)
        val envelope = linkedMapOf<String, Any?>("carga" to selada, "assinatura" to assinatura)
        return CanonicalJson.encode(envelope).toByteArray(Charsets.UTF_8)
    }

    // Hex minúsculo correto: mascara pra 0xFF e zera-preenche 2 chars por byte.
    // (o "%x"/String.format ingênuo erra com bytes >= 0x80 e < 0x10.)
    private fun ByteArray.toHexLower(): String {
        val sb = StringBuilder(size * 2)
        for (b in this) {
            val v = b.toInt() and 0xFF
            sb.append(HEX[v ushr 4])
            sb.append(HEX[v and 0x0F])
        }
        return sb.toString()
    }
}

File: ekodide/lacre.py
"""O lacre do Ekodide: assina e verifica cada mensagem com um segredo.

O segredo NUNCA cruza a rede. Quem manda assina a mensagem com ele (HMAC-SHA256);
quem recebe recalcula a assinatura e compara. É código burro e determinístico de
propósito. O lacre prova:

  - quem mandou tinha o segredo   (autenticação)
  - ninguém mexeu no caminho      (integridade)
  - a mensagem é recente          (carimbo de tempo: limita repetição a uma janela curta)

É o MESMO lacre para Wi-Fi agora e para a internet depois — ele viaja com a
mensagem, não depende do cano. O conteúdo já vai cifrado à parte (ver cofre.py,
AES-256-GCM); o lacre cuida de autenticidade/integridade/recência, não de esconder.
"""
from __future__ import annotations

import hmac
import json
import os
import time
from hashlib import sha256


class TrancaInvalida(Exception):
    """Mensagem recusada pela fechadura (assinatura, formato ou tempo)."""


# Janela do carimbo de tempo: barra repetição de mensagens antigas sem exigir
# relógios cravados no mesmo segundo. 5 min cobre folga de relógio na LAN.
JANELA_SEGUNDOS = 300


def segredo_do_ambiente() -> str:
    """Lê o segredo de EKODIDE_SEGREDO (ou OROGBO_SEGREDO, p/ compatibilidade).
    Erro claro se faltar — nunca há padrão. O CLI também sabe ler do arquivo de
    config e passar o segredo direto pra enviar()/servir()."""
    segredo = os.environ.get("EKODIDE_SEGREDO") or os.environ.get("OROGBO_SEGREDO", "")
    if not segredo:
        raise TrancaInvalida(
            "Falta o segredo: defina EKODIDE_SEGREDO (ou guarde em "
            "~/.config/ekodide/config.json com 'ekodide config')."
        )
    return segredo


def _canonico(carga: dict) -> bytes:
    """JSON estável (chaves ordenadas, sem espaços) p/ a assinatura bater dos dois lados."""
    return json.dumps(
        carga, sort_keys=True, separators=(",", ":"), ensure_ascii=False
    ).encode("utf-8")


def _assinar(corpo: bytes, segredo: str) -> str:
    return hmac.new(segredo.encode("utf-8"), corpo, sha256).hexdigest()


def empacotar(carga: dict, segredo: str, agora: float | None = None) -> bytes:
    """Carimba o tempo, assina e devolve os bytes prontos para enviar."""
    selada = {**carga, "ts": int(agora if agora is not None else time.time())}
    assinatura = _assinar(_canonico(selada), segredo)
    return _canonico({"carga": selada, "assinatura": assinatura})


def desempacotar(corpo: bytes, segredo: str, agora: float | None = None) -> dict:
    """Verifica assinatura e tempo; devolve a carga. Levanta TrancaInvalida se algo não bate."""
    try:
        envelope = json.loads(corpo)
        carga = envelope["carga"]
        assinatura = envelope["assinatura"]
    except (ValueError, KeyError, TypeError):
        raise TrancaInvalida("mensagem malformada")

    esperada = _assinar(_canonico(carga), segredo)
    # compare_digest: comparação em tempo constante (não vaza o segredo por timing).
    if not hmac.compare_digest(esperada, str(assinatura)):
        raise TrancaInvalida("assinatura não confere (segredo errado ou corpo adulterado)")

    ts = carga.get("ts")
    if not isinstance(ts, int):
        raise TrancaInvalida("sem carimbo de tempo")
    agora = agora if agora is not None else time.time()
    if abs(agora - ts) > JANELA_SEGUNDOS:
        raise TrancaInvalida("mensagem fora da janela de tempo (possível repetição)")

    return carga
